File: core/native/controller.c
#include <math.h>
#include <stdlib.h>

// System State
typedef struct {
    double position;
    double velocity;
    double integral_error;
    double last_error;
} SystemState;

SystemState sys_state = {0.0, 0.0, 0.0, 0.0};

// PID Parameters
double Kp = 2.0;
double Ki = 0.5;
double Kd = 0.1;

// MPC Parameters
#define HORIZON 10
double dt = 0.1;

// External Assembly Function
extern double blend_signals(double pid_out, double mpc_out, double alpha);

// Exported functions for Python
__declspec(dllexport) void init_system() {
    sys_state.position = 0.0;
    sys_state.velocity = 0.0;
    sys_state.integral_error = 0.0;
    sys_state.last_error = 0.0;
}

__declspec(dllexport) void set_pid_params(double p, double i, double d) {
    Kp = p;
    Ki = i;
    Kd = d;
}

__declspec(dllexport) double get_position() {
    return sys_state.position;
}

// PID Controller Implementation
double compute_pid(double target, double dt) {
    double error = target - sys_state.position;
    sys_state.integral_error += error * dt;
    double derivative = (error - sys_state.last_error) / dt;
    sys_state.last_error = error;
    
    // Anti-windup
    if (sys_state.integral_error > 10.0) sys_state.integral_error = 10.0;
    if (sys_state.integral_error < -10.0) sys_state.integral_error = -10.0;

    return (Kp * error) + (Ki * sys_state.integral_error) + (Kd * derivative);
}

// Simplified MPC Implementation (Gradient Descent approach for low complexity)
double compute_mpc(double target, double dt) {
    double predicted_pos = sys_state.position;
    double predicted_vel = sys_state.velocity;
    double best_u = 0.0;
    double min_cost = 1e9;
    
    // Simple search for optimal control action
    // In a real rigorous system we would use a QP solver, but for "less complexity" 
    // and speed in this demo, we check a range of control inputs.
    for (double u = -10.0; u <= 10.0; u += 0.5) {
        double cost = 0.0;
        double p = predicted_pos;
        double v = predicted_vel;
        
        // Simulate forward
        for (int i = 0; i < HORIZON; i++) {
            double acceleration = u - 0.1 * v; // Simple dynamics: u - friction
            v += acceleration * dt;
            p += v * dt;
            
            double error = target - p;
            cost += error * error; // Minimize squared error
        }
        
        // Add control effort penalty
        cost += 0.1 * u * u;
        
        if (cost < min_cost) {
            min_cost = cost;
            best_u = u;
        }
    }
    return best_u;
}

// Plant Physics (Second Order System)
void update_physics(double u, double dt) {
    // Dynamics: x'' = u - 0.5*x' (Damping)
    double acceleration = u - 0.5 * sys_state.velocity;
    sys_state.velocity += acceleration * dt;
    sys_state.position += sys_state.velocity * dt;
}

// Main Step Function called by Python
__declspec(dllexport) void step_system(double target, double dt, double* out_pos, double* out_u, double* out_alpha) {
    // 1. Compute PID
    double u_pid = compute_pid(target, dt);
    
    // 2. Compute MPC
    double u_mpc = compute_mpc(target, dt);
    
    // 3. Compute Alpha (Adaptive Blending)
    // Large error -> Alpha close to 1 (PID)
    // Small error -> Alpha close to 0 (MPC)
    double error = fabs(target - sys_state.position);
    double alpha = error / 5.0; 
    if (alpha > 1.0) alpha = 1.0;
    if (alpha < 0.0) alpha = 0.0;
    
    // 4. Blend using Assembly function
    double u_final = blend_signals(u_pid, u_mpc, alpha);
    
    // Clamp output
    if (u_final > 10.0) u_final = 10.0;
    if (u_final < -10.0) u_final = -10.0;
    
    // 5. Update Physics
    update_physics(u_final, dt);
    
    // 6. Return values
    *out_pos = sys_state.position;
    *out_u = u_final;
    *out_alpha = alpha;
}

File: core/controller.py
import numpy as np
from core.rl_agent import DDPGAgent

class PIDController:
    """Optimized PID Controller with Anti-Windup"""
    def __init__(self, kp=2.0, ki=0.5, kd=0.1):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.integral = 0.0
        self.last_error = 0.0
        
    def compute(self, error, dt):
        # Proportional
        P = self.kp * error
        
        # Integral with anti-windup
        self.integral += error * dt
        self.integral = np.clip(self.integral, -10.0, 10.0)
        I = self.ki * self.integral
        
        # Derivative
        derivative = (error - self.last_error) / dt if dt > 0 else 0.0
        D = self.kd * derivative
        self.last_error = error
        
        return P + I + D
    
    def reset(self):
        self.integral = 0.0
        self.last_error = 0.0


class MPCController:
    """Simplified Model Predictive Controller using Gradient Descent"""
    def __init__(self, horizon=10, dt=0.1):
        self.horizon = horizon
        self.dt = dt
        
    def compute(self, current_state, target, dt):
        """
        Simplified MPC using a grid search approach for low complexity.
        current_state: [position, velocity]
        """
        position, velocity = current_state
        best_u = 0.0
        min_cost = float('inf')
        
        # Search over control inputs
        for u in np.linspace(-10, 10, 41):  # 41 points for precision
            cost = self._predict_cost(position, velocity, u, target)
            if cost < min_cost:
                min_cost = cost
                best_u = u
                
        return best_u
    
    def _predict_cost(self, pos, vel, u, target):
        """Predict cost over horizon"""
        cost = 0.0
        p, v = pos, vel
        
        for _ in range(self.horizon):
            # Simple dynamics: acceleration = u - damping
            accel = u - 0.5 * v
            v += accel * self.dt
            p += v * self.dt
            
            # Cost: tracking error + control effort
            error = target - p
            cost += error**2 + 0.01 * u**2
            
        return cost


class PlantModel:
    """Second-Order System (e.g., DC Motor, Mass-Spring-Damper)"""
    def __init__(self):
        self.position = 0.0
        self.velocity = 0.0
        self.noise_level = 0.0
        
    def update(self, control_input, dt):
        """Update plant dynamics"""
        # Dynamics: x'' = u - 0.5*x' (damping) + disturbance
        disturbance = np.random.normal(0, self.noise_level) if self.noise_level > 0 else 0.0
        acceleration = control_input - 0.5 * self.velocity + disturbance
        self.velocity += acceleration * dt
        self.position += self.velocity * dt
        
    def reset(self):
        self.position = 0.0
        self.velocity = 0.0
    
    def set_noise(self, level):
        self.noise_level = level


class HybridController:
    """Adaptive Hybrid PID-MPC-RL Controller"""
    def __init__(self):
        self.pid = PIDController()
        self.mpc = MPCController()
        self.plant = PlantModel()
        
        # RL Agent (State: [pos, vel, target, error], Action: [u])
        self.rl_agent = DDPGAgent(state_dim=4, action_dim=1, max_action=10.0)
        self.mode = "HYBRID" # HYBRID, RL_TRAIN, RL_INFERENCE
        self.training_steps = 0
        
    def set_pid_gains(self, kp, ki, kd):
        self.pid.kp = kp
        self.pid.ki = ki
        self.pid.kd = kd
        
    def step(self, target, dt):
        """Execute one control step"""
        # 1. Calculate error
        error = target - self.plant.position
        
        # State for RL
        state = np.array([self.plant.position, self.plant.velocity, target, error])
        
        u_final = 0.0
        alpha = 0.0
        
        if self.mode == "RL_TRAIN" or self.mode == "RL_INFERENCE":
            # RL Control
            noise = 0.2 if self.mode == "RL_TRAIN" else 0.0
            action = self.rl_agent.select_action(state, noise)
            u_final = action[0]
            alpha = 0.0 # Pure RL
            
            # Train if in training mode
            if self.mode == "RL_TRAIN":
                # Calculate reward (negative squared error + penalty for effort)
                reward = -(error**2 + 0.01 * u_final**2)
                
                # Store transition (Next state is approximated for now, or we wait for next step? 
                # For simplicity in this sync loop, we'll predict next state or use a 1-step delay in a real loop.
                # Here we update plant immediately, so we can get next state.)
                
                # Update plant
                self.plant.update(u_final, dt)
                next_state = np.array([self.plant.position, self.plant.velocity, target, target - self.plant.position])
                done = False # Continuous task
                
                self.rl_agent.replay_buffer.add(state, action, reward, next_state, done)
                loss = self.rl_agent.train()
                self.training_steps += 1
                
                return self.plant.position, u_final, alpha, loss
                
        else:
            # Hybrid PID-MPC Control
            u_pid = self.pid.compute(error, dt)
            
            current_state_mpc = [self.plant.position, self.plant.velocity]
            u_mpc = self.mpc.compute(current_state_mpc, target, dt)
            
            # Adaptive Blending
            alpha = np.clip(np.abs(error) / 5.0, 0.0, 1.0)
            u_final = alpha * u_pid + (1 - alpha) * u_mpc
            u_final = np.clip(u_final, -10.0, 10.0)
            
            # Update plant
            self.plant.update(u_final, dt)
        
        return self.plant.position, u_final, alpha, 0.0
    
    def reset(self):
        self.pid.reset()
        self.plant.reset()
    
    def set_mode(self, mode):
        self.mode = mode
    
    def set_noise(self, level):
        self.plant.set_noise(level)

